Mathlib/CategoryTheory/Limits/Types/Multicoequalizer.lean

```diff
@@ -0,0 +1,95 @@
+/-
+Copyright (c) 2025 Joël Riou. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joël Riou
+-/
+import Mathlib.CategoryTheory.Limits.Preserves.Shapes.Multiequalizer
+import Mathlib.CategoryTheory.Limits.Types.Colimits
+import Mathlib.Data.Set.BooleanAlgebra
+import Mathlib.Data.Set.FunctorToTypes
+import Mathlib.Order.CompleteLattice.MulticoequalizerDiagram
+
+/-!
+# Multicoequalizers in the category of types
+
+Given `J : MultispanShape`, `d : MultispanIndex J (Type u)` and
+`c : d.multispan.CoconeTypes`, we obtain a lemma `isMulticoequalizer_iff`
+which gives a criteria for `c` to be a colimit (i.e. a multicoequalizer):
+it restates in a more explicit manner the injectivity and surjectivity
+conditions for the map `d.multispan.descColimitType c : d.multispan.ColimitType → c.pt`.
+
+We deduce a definition `Set.isColimitOfMulticoequalizerDiagram` which shows
+that given `X : Type u`, a `MulticoequalizerDiagram` in `Set X` gives
+a multicoequalizer in the category of types.
+
+-/
+
+universe w w' u
+
+namespace CategoryTheory.Functor.CoconeTypes
+
+open Limits
+
+lemma isMulticoequalizer_iff {J : MultispanShape.{w, w'}} {d : MultispanIndex J (Type u)}
+  (c : d.multispan.CoconeTypes) :
+    c.IsColimit ↔
+      (∀ (i₁ i₂ : J.R) (x₁ : d.right i₁) (x₂ : d.right i₂)
+        (_ : c.ι (.right i₁) x₁ = c.ι (.right i₂) x₂),
+          d.multispan.ιColimitType (.right i₁) x₁ = d.multispan.ιColimitType (.right i₂) x₂) ∧
+      (∀ (x : c.pt), ∃ (i : J.R) (a : d.right i), c.ι (.right i) a = x) := by
+  have (x : d.multispan.ColimitType) :
+      ∃ (i : J.R) (a : d.right i), d.multispan.ιColimitType (.right i) a = x := by
+    obtain ⟨(l | r), z, rfl⟩ := d.multispan.ιColimitType_jointly_surjective x
+    · exact ⟨J.fst l, d.multispan.map (WalkingMultispan.Hom.fst l) z,
+        by rw [ιColimitType_map]⟩
+    · exact ⟨r, z, by simp⟩
+  constructor
+  · intro hc
+    refine ⟨fun i₁ i₂ x₁ x₂ h ↦ ?_, ?_⟩
+    · simp only [← descColimitType_ιColimitType_apply] at h
+      exact hc.bijective.1 h
+    · intro x
+      obtain ⟨y, rfl⟩ := hc.bijective.2 x
+      obtain ⟨i, z, rfl⟩ := this y
+      exact ⟨i, z, by simp⟩
+  · rintro ⟨h₁, h₂⟩
+    refine ⟨fun x₁ x₂ h ↦ ?_, fun x ↦ ?_⟩
+    · obtain ⟨i₁, a₁, rfl⟩ := this x₁
+      obtain ⟨i₂, a₂, rfl⟩ := this x₂
+      exact h₁ _ _ _ _ h
+    · obtain ⟨i, y, rfl⟩ := h₂ x
+      exact ⟨d.multispan.ιColimitType (.right i) y, rfl⟩
+
+end CategoryTheory.Functor.CoconeTypes
+
+open CompleteLattice CategoryTheory Limits
+
+namespace Set
```

Comment:
It seems to make more sense to me to put this in `CategoryTheory.Limits.Types`, mimicking `isColimit_iff_coconeTypesIsColimit`; I don't think we have much category stuff in `Set`.